android/build.gradle

```diff
@@ -1,3 +1,38 @@
+import groovy.json.JsonSlurper
+
+def safeExtGet(prop, fallback) {
+    rootProject.ext.has(prop) ? rootProject.ext.get(prop) : fallback
+}
+
+def warnNoAutoInit() {
+    logger.warn("Couldn't auto-initialize the Heap library; manually initialize it by following the instructions at https://docs.heap.io/docs/react-native.")
+}
+
+def warnDisabled() {
+    logger.warn("Not auto-initializing the Heap library because it was manually disabled; set `enabled` to `true` in `heap.json` to enable auto-initialization.")
+}
+
+ext.HEAP_CONFIG_FILENAME = 'heap.config.json'
+ext.HEAP_APP_ID_FIELD_NAME_DEV = 'HEAP_APP_ID_DEV'
+ext.HEAP_APP_ID_FIELD_NAME_PROD = 'HEAP_APP_ID_PROD'
+
+def heapAppId(buildType) {
+    def file = rootProject.file("../${HEAP_CONFIG_FILENAME}")
+    if (!file.exists()) { return warnNoAutoInit() }
+
+    def jsonSlurper = new JsonSlurper()
+    def data = jsonSlurper.parseText(file.text.trim())[buildType]
+    def defaultData = jsonSlurper.parseText(file.text.trim())["default"]
+
+    // KLUDGE: Can't use the elvis operator because we want to treat `null` and `false` differently.
+    def autoInit = data?.heapAutoInit
+    if (autoInit == null) { autoInit = defaultData?.heapAutoInit }
```

Comment:
It's really not clear to me what's going on here and the rest of the function ... now granted, it's been a while since I've gotten groovy, but I suspect there's something you could do to reorganize the conditions a bit, for clarity.  Also, is the repeated `autoInit == null` a mistake?

Reply:
Suggestions? This seemed fairly readable to me; the intention essentially is: if `heapAutoInit` for the current build type is set, use it; if not (it's null), and `heapAutoInit` for the `default` build type is set, use it; if not (it's null too), just use `true`.

Reply:
So I think I usually think of it upside down from what you've got.  Here's what I came up with, and I think they're equivalent (let me know if not!)
```groovy
    def autoInit = true // Default value
    if (data?.heapAutoInit != null) {
      autoInit = data?.heapAutoInit
    } else if (defaultData?.heapAutoInit != null) {
      autoInit = defaultData.heapAutoInit
    }
```
... definitely a bit more verbose, but I think a bit clearer.  It just _looks_ like an easily-made error from 5000 ft to have that same condition repeated twice - you have to actually grok the rest of the line to realize that the variable in the condition got modified in between those checks!

Reply:
Could even do something like (completely spitballing, haven't tried it):
```groovy
    def autoInit = true // Default value
    def configValue = null
    if ((configValue = data?.heapAutoInit) != null) {
      autoInit = configValue
    } else if ((configValue = defaultData?.heapAutoInit) != null) {
      autoInit = configValue
    }
```

Reply:
Ack. These versions seem _less_ readable to me, but I have been looking at this for a while, so your impression is likely a lot more accurate than mine. 🙂